scheduler-driver/src/main/kotlin/gov/nasa/jpl/aerie/scheduler/plan/InMemoryEditablePlan.kt

```diff
@@ -11,14 +11,41 @@ import gov.nasa.ammos.aerie.procedural.timeline.payloads.activities.AnyDirective
 import gov.nasa.ammos.aerie.procedural.timeline.payloads.activities.Directive
 import gov.nasa.ammos.aerie.procedural.timeline.payloads.activities.DirectiveStart
 import gov.nasa.ammos.aerie.procedural.timeline.plan.Plan
+import gov.nasa.ammos.aerie.procedural.timeline.plan.SimulationResults
 import gov.nasa.jpl.aerie.merlin.protocol.types.DurationType
 import gov.nasa.jpl.aerie.scheduler.DirectiveIdGenerator
 import gov.nasa.jpl.aerie.scheduler.model.*
 import gov.nasa.jpl.aerie.types.ActivityDirectiveId
+import java.lang.ref.WeakReference
 import java.time.Instant
 import kotlin.jvm.optionals.getOrNull
 import gov.nasa.ammos.aerie.procedural.timeline.plan.SimulationResults as TimelineSimResults
 
+/**
+ * An implementation of [EditablePlan] that stores the plan in memory for use in the internal scheduler.
+ *
+ * ## Staleness checking
+ *
+ * The editable plan instance keeps track of sim results that it has produced using weak references, and can dynamically
+ * update their staleness if the plan is changed after it was simulated. The process is this:
+ *
+ * 1. [InMemoryEditablePlan] has a set of weak references to simulation results objects that are currently up-to-date.
+ *    I used weak references because if the user can't access it anymore, staleness doesn't matter and we might as well
+ *    let it get gc'ed.
+ * 2. When the user gets simulation results, either through simulation or by getting the latest, it always checks for
+ *    plan equality between the returned results and the current plan, even if we just simulated. If it is up-to-date, a
+ *    weak ref is added to the set.
```

Comment:
Does this mean that the user could repeatedly add to this set by, ie, getting the latest results in a loop?

Reply:
I actually don't know, it depends on if the simulation facade returns the exact same pointer, rather than equivalent data, each time. And I don't know which one it does.

But since they are weak references, they will not prevent the sim results from being garbage collected when the user code loses all strong references to it. So it doesn't actually matter; the set could contain a thousand weak references to the same Sim Results instance, or it could contain a thousand references to different instances with the same data. But since they are all weak references, the memory usage will be the same as before this PR.

Reply:
> I actually don't know, it depends on if the simulation facade returns the exact same pointer, rather than equivalent data, when the user simulates the same plan multiple times. And I don't know which one it does.

While I see logic to avoid a resim if the plan matches the initial results, I see no such logic for anything past that. It looks like at best, the facade will recalculate a new Results object with the same data, and at worst it will resim the portion of the plan between the Checkpoint and the end time (Which is a regression. Near as I can tell, the reason we aren't noticing this regression is because our test for this behavior (`noNeedForResimulation`) doesn't actually check that any of the objects are the same, just that the results are _equivalent_. But the engine still goes all the way into the sim loop and just happens to stop immediately because the checkpoint was already after the end of the requested horizon).

> But since they are all weak references, the memory usage will be the same as before this PR.

Understood, although could you quickly go over why we need a _Set_ of up to date results for a commit? Is it because we don't know _which_ sim results the user still has a strong reference to?

Reply:
This PR doesn't change any of the behavior relating to when the sim facade does or doesn't simulate, or whether the sim facade returns a cached results object if simulated twice. That is up to the sim facade and its behavior should be exactly the same after this change.

> why we need a Set of up to date results for a commit?

The user will probably only simulate at most once for any given plan state, but we don't know that for sure. The user can simulate as much as they want, and the sim facade can return whatever it wants, and this PR treats both of them as black boxes. So there might be multiple equivalent sim results objects floating around. Its unlikely but I wanted to be safe.